File: Sorts & Search & Shuffling Algorithms/Elementary Sorts/Pigeonhole Sort/Pigeonhole.py
class Pigeonhole:
    @staticmethod
    def sort(a):
        my_min = min(a)
        my_max = max(a)
        size = my_max - my_min + 1
        holes = [0] * size
        for x in a:
            assert type(x) is int, "integers only please"
            holes[x - my_min] += 1
        i = 0
        for count in range(size):
            while holes[count] > 0:
                holes[count] -= 1
                a[i] = count + my_min
                i += 1

File: Sorts & Search & Shuffling Algorithms/Elementary Sorts/Pigeonhole Sort/Pigeonhole.cpp
#include <vector>
#include <cstdlib>
#include <ctime>
using namespace std;
class Pigeonhole {
public:
    static void Sort(vector<int> &arr) {
        int min = arr[0], max = arr[0];
        int n=arr.size();
        for (int i = 1; i < n; i++){
            if (arr[i] < min)
                min = arr[i];
            if (arr[i] > max)
                max = arr[i];
        }
        int range = max - min + 1;
        vector<int> holes[range];
        for (int i = 0; i < n; i++)
            holes[arr[i]-min].push_back(arr[i]);
        int index = 0;
        for (int i = 0; i < range; i++){
            vector<int>::iterator it;
            for (it = holes[i].begin(); it != holes[i].end(); ++it)
                arr[index++]  = *it;
        }
    }
};